File: flink_scala/src/main/scala/github/yahuili1128/streamOperators/ReduceDemo.scala
package github.yahuili1128.streamOperators

import github.yahuili1128.connector.ScalaSourceKafka010
import github.yahuili1128.connector.ScalaSourceKafka010.MockUp
import org.apache.flink.api.common.functions.ReduceFunction
import org.apache.flink.streaming.api.scala._

/**
  * @Description : flink reduce scala
  * @Author : LiYahui
  * @Date : 2019-07-15 18:00
  * @Version : V1.0
  */
object ReduceDemo {
	def main(args: Array[String]): Unit = {
		val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
		val mockup: DataStream[ScalaSourceKafka010.MockUp] = ScalaSourceKafka010.getMockUpFromKafka(env)
		val resDS: DataStream[ScalaSourceKafka010.MockUp] = mockup
				.keyBy(_.gender)
				.reduce(new ReduceFunction[ScalaSourceKafka010.MockUp] {
					override def reduce(t: ScalaSourceKafka010.MockUp, t1: ScalaSourceKafka010.MockUp): ScalaSourceKafka010.MockUp = {
						MockUp(t.name + "--" + t1.name, t.gender, t.timestamp, t.age + t1.age)
					}
				})
		resDS.print().setParallelism(1)

		env.execute(s"${this.getClass.getSimpleName}")
	}
}

File: flink_java/src/main/java/github/yahuili1128/streamOperators/ReduceDemo.java
package github.yahuili1128.streamOperators;

import github.yahuili1128.pojo.MockUpModel;
import org.apache.flink.api.common.functions.ReduceFunction;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

import static github.yahuili1128.connector.SourceKafka010.getMockUpkafka010;

/**
 * @Description :flink java reduce
 * @Author : LiYahui
 * @Date : 2019-06-26 11:27
 * @Version : V1.0
 */
public class ReduceDemo {
	public static void main(String[] args) throws Exception {
		final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
		SingleOutputStreamOperator<MockUpModel> kafka010Data = getMockUpkafka010(env);
		//		lambda +java
		kafka010Data.keyBy(line -> line.gender).reduce(new ReduceFunction<MockUpModel>() {
			@Override
			public MockUpModel reduce(MockUpModel value1, MockUpModel value2) throws Exception {
				MockUpModel mockUpModel = new MockUpModel();
				mockUpModel.name = value1.name + "--" + value2.name;
				mockUpModel.gender = value1.gender;
				mockUpModel.age = (value1.age + value2.age) / 2;
				return mockUpModel;

			}
		}).print().setParallelism(1);

		env.execute("flink kafka010 demo");
	}
}
